clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -57,238 +55,51 @@
-        }
-
-        public Errors error() {
-            return error;
-        }
-
-        public long highWatermark() {
-            return partitionResponse.highWatermark();
-        }
-
-        public long lastStableOffset() {
-            return partitionResponse.lastStableOffset();
-        }
-
-        public long logStartOffset() {
-            return partitionResponse.logStartOffset();
-        }
-
-        public Optional<Integer> preferredReadReplica() {
-            return preferredReplica;
-        }
-
-        public List<AbortedTransaction> abortedTransactions() {
-            return abortedTransactions;
-        }
-
-        public Optional<FetchResponseData.EpochEndOffset> divergingEpoch() {
-            FetchResponseData.EpochEndOffset epochEndOffset = partitionResponse.divergingEpoch();
-            if (epochEndOffset.epoch() < 0) {
-                return Optional.empty();
-            } else {
-                return Optional.of(epochEndOffset);
-            }
-        }
-
-        @SuppressWarnings("unchecked")
-        public T records() {
-            return (T) partitionResponse.recordSet();
-        }
-    }
-
-    /**
-     * From version 3 or later, the entries in `responseData` should be in the same order as the entries in
-     * `FetchRequest.fetchData`.
-     *
-     * @param error             The top-level error code.
-     * @param responseData      The fetched data grouped by partition.
-     * @param throttleTimeMs    The time in milliseconds that the response was throttled
-     * @param sessionId         The fetch session id.
-     */
     public FetchResponse(Errors error,
-                         LinkedHashMap<TopicPartition, PartitionData<T>> responseData,
                          int throttleTimeMs,
-                         int sessionId) {
-        super(ApiKeys.FETCH);
-        this.data = toMessage(throttleTimeMs, error, responseData.entrySet().iterator(), sessionId);
-        this.responseDataMap = responseData;
+                         int sessionId,
+                         LinkedHashMap<TopicPartition, FetchResponseData.FetchablePartitionResponse> responseData) {
+        this(new FetchResponseData()
+            .setSessionId(sessionId)
+            .setErrorCode(error.code())
+            .setThrottleTimeMs(throttleTimeMs)
+            .setResponses(responseData.entrySet().stream().map(entry -> new FetchResponseData.FetchableTopicResponse()
+                .setTopic(entry.getKey().topic())
+                .setPartitionResponses(Collections.singletonList(entry.getValue().setPartition(entry.getKey().partition()))))
+                .collect(Collectors.toList())));
     }
 
     public FetchResponse(FetchResponseData fetchResponseData) {
         super(ApiKeys.FETCH);
         this.data = fetchResponseData;
-        this.responseDataMap = toResponseDataMap(fetchResponseData);
+        this.responseData = new LinkedHashMap<>();
```

Comment:
This isn't needed when we return a fetch from the broker, right? If this is true, can we remove it from the fetch response and built it on the client when needed?